File: src/test/java/com/pschlup/ta/indicators/AverageTrueRangeTest.java
package com.pschlup.ta.indicators;

import com.pschlup.ta.timeseries.Bar;
import com.pschlup.ta.timeseries.CsvBarSource;
import com.pschlup.ta.timeseries.TimeSeries;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.List;

import static com.google.common.truth.Truth.assertThat;

class AverageTrueRangeTest {

  private TimeSeries timeSeries;

  @BeforeEach
  void setup() throws IOException {
    List<Bar> bars = CsvBarSource.readCsvData(
        AverageTrueRangeTest.class.getResourceAsStream("/test_chart_data.csv"));
    this.timeSeries = TimeSeries.of(bars);
  }

  @Test
  void shouldComputeValue() {
    AverageTrueRange atr = AverageTrueRange.of(timeSeries).withLength(14);
    assertThat(atr.valueAt(0))
        .isWithin(0.0001).of(49.49144);
    assertThat(atr.valueAt(1))
        .isWithin(0.0001).of(49.43149);
  }
}

File: src/test/kotlin/com/pschlup/ta/indicators/AverageTrueRangeTest.kt
package com.pschlup.ta.indicators

import com.google.common.truth.Truth.assertThat
import com.pschlup.ta.timeseries.CsvBarSource.readCsvData
import com.pschlup.ta.timeseries.TimeSeries
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class AverageTrueRangeTest {
  private lateinit var timeSeries: TimeSeries

  @BeforeEach
  fun setup() {
    val bars = readCsvData(
      javaClass.getResourceAsStream("/test_chart_data.csv") ?: error("CSV not found")
    )
    timeSeries = TimeSeries.of(bars)
  }

  @Test
  fun `should compute value`() {
    val atr = timeSeries.averageTrueRange(14)

    // Double-check values with another source
    assertThat(atr.valueAt(0))
      .isWithin(0.0001).of(47.8222)
    assertThat(atr.valueAt(1))
      .isWithin(0.0001).of(48.4050)
  }
}
